contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
 
 ### Kubernetes API Changes
 
-#### Chosen Solution: "exec++"
+#### Chosen Solution: POST to an Existing Pod
 
-We will extend `v1.Pod`'s `/exec` subresource to support "executing" container
-images. The current `/exec` endpoint must implement `GET` to support streaming
-for all clients. We don't want to encode a (potentially large) `v1.Container` as
-an HTTP parameter, so we must extend `v1.PodExecOptions` with the specific
-fields required for creating a Debug Container:
+We're modifying an existing pod, so this fits as a subresource of the target
+pod. Since we're required to use `v1.Container`, the API should use `POST` or
+`PUT` to avoid encoding this relatively large object into a URL query string. If
+we `POST` or `PUT` then we cannot make this a streaming call if we want to
+continue supporting web socket clients.
+
+We can either using an existing subresource like `/exec` and have the call
+stream or not based on `PodExecOptions`, or we can create a new subresource that
+never streams. For the sake of consistency, we will create a new subresource
+named `/debug`. This has the added benefit of being able to entirely hide
+interface behind a feature flag by conditionally registering the new
+subresource.
+
+As `v1.Container` lacks type and object metadata, we must wrap it in a new
+object:
+
+```
+// EphemeralContainer describes a container to attach to a running pod for troubleshooting.
+type EphemeralContainer struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        // Spec describes the Ephemeral Container to be created.
+        Spec Container `json:"spec" ...`
+}
+```
+
+Note that Debug Containers are not regular containers and should not be used to
+build services. They have no guarantees and most of the fields of `v1.Container`
+will not be allowed for Debug Containers. A request will be rejected if any
+field is set other than the following whitelisted fields: `Name`, `Image`,
+`Command`, `Args`, `WorkingDir`, `Env`, `EnvFrom`, `ImagePullPolicy`,
+`SecurityContext`. `TTY` and `Stdin` are always enabled for Debug Containers and
+will be ignored.
+
+The new `/debug` subresource allows the following:
+
+1.  A `POST` of a `PodDebugContainer` to
+    `/api/v1/namespaces/$NS/pods/$POD_NAME/debug` to create a Debug Container
+    running in pod `$POD_NAME`.
+1.  A `DELETE` of `/api/v1/namespaces/$NS/pods/$POD_NAME/debug/$NAME` will stop
+    the Debug Container `$NAME` in pod `$POD_NAME`.
+
+Once created, a client would attach to the console of a debug container using
+the existing attach endpoint, `/api/v1/namespaces/$NS/pods/$POD_NAME/attach`.
+Note that any output of the new container between its creation and subsequent
+attach will not be replayed and can only be viewed using `kubectl log`.
+
```

Comment:
Will the ephemeral container appear in the response when I GET the pod?

Reply:
only the status of it, not the spec

Reply:
That's fine.